WIP 1714 fluent react

fzzzy opened the pull request.

fzzzy committed:
Merge remote-tracking branch 'mozilla/master'

fzzzy committed:
Merge branch 'master' of https://github.com/mozilla/testpilot

fzzzy committed:
Merge remote-tracking branch 'mozilla/master'

fzzzy committed:
Step 1 get it working

fzzzy committed:
Remove console.log

fzzzy committed:
Switch View over to fluent-react

fzzzy committed:
Switch ErrorPage to fluent-react

fzzzy committed:
Convert ExperimentPage to fluent-react

fzzzy committed:
Convert HomePageNoAddon to fluent-react

fzzzy committed:
Convert HomePageWithAddon to fluent-react

fzzzy committed:
Convert NotFoundPage to fluent-react

fzzzy committed:
Convert OnboardingPage to fluent-react.

fzzzy committed:
Convert RestartPage to fluent-react.

fzzzy committed:
Convert RetirePage to fluent-react.

fzzzy committed:
Convert SharePage to fluent-react.

fzzzy committed:
Convert EmailDialog to fluent-react.

fzzzy committed:
Convert ExperimentDisableDialog to fluent-react

fzzzy committed:
Convert ExperimentEolDialog to fluent-react

fzzzy committed:
Convert ExperimentPreFeedbackDialog to fluent-react.

fzzzy committed:
Convert ExperimentRowCard to fluent-react.

fzzzy committed:
Convert ExperimentTourDialog to fluent-react

fzzzy committed:
Convert Footer to fluent-react

fzzzy committed:
Convert Header to fluent-react

fzzzy committed:
Convert MainInstallButton to fluent-react

fzzzy committed:
Convert NewsletterFooter to fluent-react.

fzzzy committed:
Convert NewsletterForm to fluent-react.

fzzzy committed:
Convert PastExperiments to fluent-react.

fzzzy committed:
Convert RetireConfirmationDialog to fluent-react.

fzzzy committed:
Convert Settings to fluent-react.

fzzzy committed:
Convert UpdateList to fluent-react.

fzzzy committed:
Convert Warning to fluent-react

fzzzy committed:
Fix typos.

fzzzy committed:
Move the key attributes to the right element.

fzzzy committed:
Remove trailing whitespace

fzzzy committed:
Don't try to JSON.parse undefined

fzzzy committed:
Remove copy-pasted junk

fzzzy committed:
Fix the type of titleL10nArgs

fzzzy committed:
Quote apostrophe

fzzzy committed:
Annotate the correct return type

pdehaan commented on frontend/src/app/components/EmailDialog.js:
```diff
@@ -49,11 +50,15 @@ export default class EmailDialog extends React.Component {
         <div className="modal-content modal-form centered">
-          <p data-l10n-id="emailOptInMessage" className="">Find out about new experiments and see test results for experiments you&apos;ve tried.</p>
+          <Localized id="emailOptInMessage">
+            <p className="">Find out about new experiments and see test results for experiments you&apos;ve tried.</p>
```
Does this need the empty `className` attribute?

pdehaan commented on frontend/src/app/components/EmailDialog.js:
```diff
@@ -68,15 +73,21 @@ export default class EmailDialog extends React.Component {
         <div className="modal-actions">
-          <button id="email-success-continue" onClick={e => this.continue(e)} className="button default large" data-l10n-id="emailOptInConfirmationClose">On to the experiments...</button>
+          <Localized id="emailOptInConfirmationClose">
+            <button id="email-success-continue" onClick={e => this.continue(e)} className="button default large">On to the experiments...</button>
```
I think @flodolo generally recommends the `&hellip;` "&hellip;" over the 3 separate dots "...".

stasm commented on frontend/src/app/components/MainInstallButton.js:
```diff
@@ -88,16 +102,22 @@ export default class MainInstallButton extends React.Component {
             ) : (
-              <span className="parens" data-l10n-id="landingUpgradeDesc2" data-l10n-args={JSON.stringify({ version: config.minFirefoxVersion })}>Test Pilot requires Firefox { config.minFirefoxVersion } or higher.</span>
+              <Localized id="landingUpgradeDesc2" $version={config.minFirefoxVersion}>
+                <span className="parens">Test Pilot requires Firefox { config.minFirefoxVersion } or higher.</span>
```
There are two ways to approach what should go between into the `<span>` here and I'd like to learn your take on which one is more desirable.

The first one is what you did here: put the valid JSX into the `<span>`.  It can serve as a fallback for when the `landingUpgradeDesc2` translation is missing. The drawback is that React will create the virtual DOM nodes for the content which `Localized` will then remove in favor of the translated string.

The other approach is to use valid FTL syntax as `span`'s content:

```
<Localized id="landingUpgradeDesc2" $version={config.minFirefoxVersion}>
  <span className="parens">
    {'Test Pilot requires Firefox { $version } or higher.'}
  </span>
</Localized>
```

This has the benefit of being exactly the string that needs to be translated which in the future could be used for extracting English strings into an FTL file. On the other hand, it doesn't work as well as the first approach when it comes to fallback.  it's just a static string and in particular, `$version` wouldn't get replaced if `landingUpgradeDesc2` is missing from all languages (English included).

fzzzy replied:
I don't have much of an opinion on this. My gut feeling is to go with the valid jsx. Even though the nodes will just be tossed, it doesn't seem like that big of an issue.

stasm commented on frontend/src/app/components/NewsletterForm.js:
```diff
@@ -75,36 +76,44 @@ export default class NewsletterForm extends React.Component {
+        <Localized id="newsletterFormPrivacyNotice">
+          <span>
+            I&apos;m okay with Mozilla handling by info as explained in
+            <a target="_blank" rel="noopener noreferrer"
```
It's currently not possible to have nested elements inside of `Localized` components. Instead you may try passing the `<a>` element as a `$`-prefixed argument to `Localized`. See the bottom of https://github.com/projectfluent/fluent.js/wiki/Localized for an example of this approach.

Adding support for nested elements is definitely planned, although I haven't made much research into it and don't know how hard it's going to be.

stasm commented on frontend/src/app/containers/App.js:
```diff
@@ -100,14 +104,51 @@ class App extends Component {
+      }
+    }
+
+    Promise.all(navigator.languages.map(language =>
```
`navigator.languages` might include languages that Test Pilot doesn't support. You can use [fluent-langneg](https://github.com/projectfluent/fluent.js/tree/master/fluent-langneg) to make sure you only fetch files that are known to exist.

fzzzy replied:
Awesome! Thanks!

stasm commented on frontend/src/app/containers/App.js:
```diff
@@ -100,14 +104,51 @@ class App extends Component {
+          fetch(`/static/locales/${language}/experiments.ftl`).then(response => addLang(language, response))
+        ]
+      )
+    )).then(() => this.props.setLocalizations(langs));
```
You could possibly store the entire `messages` iterator in the state, instead of raw localization files, which would make the code in `render` a bit simpler.  You also wouldn't need the `addLang` function; you can use `cx.addMessages(data);` multiple times on a single `MessageContext` instance.

fzzzy replied:
I've decided to do this in another PR since the current behavior works and this PR is getting huge. I opened a bug for it: https://github.com/mozilla/testpilot/issues/2609

stasm commented on frontend/src/app/containers/ExperimentPage.js:
```diff
@@ -146,12 +147,16 @@ export class ExperimentDetail extends React.Component {
+          </Localized>
+          <Localized id="incompatibleSubheader">
+            <p>
+              We recommend <a href={helpUrl}>disabling these add-ons</a> before activating this experiment:
```
Similar to my comment above, the workaround to `fluent-react` not supporting nested elements is to pass `<a>` as an argument to `Localized`. See the bottom of https://github.com/projectfluent/fluent.js/wiki/Localized.

fzzzy replied:
Great, thanks a lot. However what if the contents of the anchor needs to be localized too?

stasm replied:
You can pass `<Localized>` elements as arguments, too :) It's not ideal, but it's also a common practice in other React localization libraries from what I've seen.

fzzzy replied:
Ok, I wondered if that would work. 👍

fzzzy replied:
I'm making some progress removing the html from localized strings. The described technique works fine, and having nested Localized components works fine too.

stasm commented on frontend/src/app/containers/ExperimentPage.js:
```diff
@@ -307,13 +319,19 @@ export class ExperimentDetail extends React.Component {
-              {(statusType === 'enabled') && <span data-l10n-id="isEnabledStatusMessage" data-l10n-args={JSON.stringify({ title })}><span></span></span>}
-              {(statusType === 'error') && <span data-l10n-id="installErrorMessage" data-l10n-args={JSON.stringify({ title })}><span></span></span>}
+              {(statusType === 'enabled') && <Localized id="isEnabledStatusMessage" $title={ title }>
+                <span><span></span></span>
```
What is the reason for having two `spans` here? You might need to put the outer one outside of the `statusType` check.

fzzzy replied:
I'm not sure, probably something having to do with CSS. I'll ask John about it.

stasm commented on frontend/src/app/reducers/localizations.js:
```diff
@@ -0,0 +1,37 @@
+  return action.payload;
+}
+
+export default function experimentsReducer(
```
Rename this to `localizationsReducer`?

fzzzy replied:
Good catch :)

stasm reviewed (commented):
Whoa, that's one huge PR.  It looks great too. Let me know if you have any questions about `fluent-react`.

fzzzy committed:
Rename

fzzzy committed:
Remove empty className

fzzzy committed:
Use &hellip;

fzzzy committed:
Remove weird double nested span.

stasm committed:
WIP Fix test failures caused by fluent-react

fluent-react 0.4.1 relaxes the restriction about Localized components being
only valid inside of LocalizationProvider.  The tests now pass but React
reports the following warning:

    Warning: Failed context type: The l10n context field must be an instance
    of ReactLocalization.
        in Localized (created by NewsletterForm)
        in label (created by NewsletterForm)
        in form (created by NewsletterForm)
        in NewsletterForm (created by Unknown)
        in Unknown

The warning could be silenced by temporarily muting console.warn, or by
implementing a wrapper around Enzyme's mount which provides a fake
ReactLocalization in the context.

Here's what the wrapper could look like:

    https://github.com/stasm/fluent.js/blob/fluent-enzyme/fluent-enzyme/src/index.js#L35-L49

I also factored the findLocalizedById function out of tests and into the
util.js module.

There are 7 skipped tests. I wasn't sure what their result HTML should be.

fzzzy committed:
Remove some of the html embedded in strings.

fzzzy committed:
Remove embedded html from the rest of app.ftl strings

Commit:
Merge pull request #4 from stasm/fluent-react-tests

WIP Fix test failures caused by fluent-react

fzzzy committed:
Use fluent-react/compat so we don't include the un-babelified version in vendor.js

fzzzy committed:
Implement LocalizedHtml component and use it to parse the strings on the ExperimentPage that contain html

fzzzy committed:
Fix lint

fzzzy committed:
Use fluent-langneg

fzzzy committed:
Add fluent-react to flow-typed
